File: love_and_care/love_and_care/lib/domain/auth/user_model.dart
import 'package:equatable/equatable.dart';

class User extends Equatable {
  final String userId;
  final String username;
  final String email;
  final String role;
  final String token;
  final String phoneNumber;
  final String address;
  final String interests;
  final String about;
  final String skills;

  // Add other fields as needed

  User({
    required this.userId,
    required this.username,
    required this.email,
    required this.role,
    required this.token,
    required this.phoneNumber,
    required this.address,
    required this.interests,
    required this.about,
    required this.skills,

    // Initialize other fields as needed
  });

  factory User.fromJson(Map<String, dynamic> json) {
    return User(
      userId: json['id'] as String,
      username: json['username'] as String,
      token: json['token'] as String,
      email: json['email'] as String,
      phoneNumber: json['phoneNumber'] as String,
      role: json['role'] as String,
      address: json['address'] as String,
      interests: json['interests'] as String,
      about: json['about'] as String,
      skills: json['skills'] as String,

      // Initialize other fields as needed
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': userId,
      'username': username,
      'token': token,
      'email': email,
      'role': role,
      'skills': skills,
      'about': about,
      'interests': interests,
      'address': address,
      'phoneNumber': phoneNumber,

      // Add other fields as needed
    };
  }

  @override
  List<Object?> get props => [
        userId,
        username,
        token,
        email,
        role,
        skills,
        interests,
        phoneNumber,
        address,
        about
      ];
}

File: love_and_care/server/models/user_model.js
const mongoose = require("mongoose");
const Joi = require("joi");

const config = require("config");
const jwt = require("jsonwebtoken");

const userSchema = new mongoose.Schema({
  username: {
    type: String,
    required: true,
    maxlength: 50,
    unique: true,
  },
  role: {
    type: String,
    required: true,
    maxlength: 50,
  },
  email: {
    type: String,
    required: true,
    maxlength: 255,
    unique: true,
  },
  password: {
    type: String,
    required: true,
    maxlength: 1024,
  },
  phoneNumber: {
    type: String,
    maxlength: 15,
  },
  address: {
    type: String,
    maxlength: 255,
  },
  skills: {
    type: String,
    maxlength: 500,
  },
  interests: {
    type: String,
    maxlength: 255,
  },
  about: {
    type: String,
    maxlength: 1000,
  },
  default: false,
});

userSchema.methods.generateAuthToken = function () {
  const token = jwt.sign(
    { _id: this._id, role: this.role, username: this.username },
    config.get("jwtPrivateKey")
  );
  return token;
};

const User = mongoose.model("User", userSchema);

function validateUserRegister(user) {
  const schema = Joi.object({
    username: Joi.string().max(50).required(),
    role: Joi.string().max(50).required(),
    email: Joi.string().max(255).required().email(),
    password: Joi.string().max(1024).required(),
  });
  return schema.validate(user);
}
function validateUserLogging(user) {
  const schema = Joi.object({
    email: Joi.string().max(255).required().email(),
    password: Joi.string().max(1024).required(),
  });
  return schema.validate(user);
}

exports.User = User;
exports.validateUserRegister = validateUserRegister;
exports.validateUserLogging = validateUserLogging;
